### PR TITLE
Move elements of Mongo into this lab

### LINKED ISSUE #4: Fix a "pending" Karma test
Current the "should not allow a name to contain a symbol" karma test in `add-user.component.spec.ts` doesn't seem to actually be running and shows up as "pending" in the test runner. It also triggers a warning in IntelliJ about using a "comma construction" which is deemed as Bad. I think the problem in both cases is the fact that there's a close paren right after the expectation string in the first line, which I think ends the `it` clause. The body is then ignored and not run (I think).

We should figure out what's going on here and fix this.

### LINKED ISSUE #8: Move elements of Mongo into this lab
There are many Mongo elements that will need to be added to this repository (either from the old lab 4 [https://github.com/UMM-CSci-3601/3601-lab4_mongo-pre-S20](https://github.com/UMM-CSci-3601/3601-lab4_mongo-pre-S20) or from your imaginations...)

### LINKED ISSUE #10: Properly unsubscribe from getUserById in user profile
The subscription created here should unsubscribed when a different ID is passed in and we subscribe to a different user. In practice since we are subscribing to a simple HTTP call it will never omit past the first time but it is good practice since in principle it could emit again.
https://github.com/UMM-CSci-3601/3601-lab4-mongo/blob/1e5b5a48fc1f3bb4cd54aa55154222d3157282be/client/src/app/users/user-profile.component.ts#L22-L25

### PR DESCRIPTION
**Pull request by wallerli**

### CODE REVIEW

**Comment by wallerli** [comment]
- Currently all server's functions have been transferred to use MongoDB and Jackson.

- The add-new-user function has not been justified yet due to lack of test. And server tests have not been updated yet and are currently disabled.

- We also need to setup the mongoDB tasks in build.gradle and Update e2e.yaml to let GH actions run the mongoDB together with the server.

**Comment by floogulinc** [comment]
Oops I hit ready for review by accident, please consider it not ready yet.

File: server/src/main/java/umm3601/user/UserController.java [hunk 1]
```diff
@@ -29,24 +51,100 @@ public UserController(UserDatabase userDatabase) {
...
+  }
+
+  /**
+   * Get a JSON response with a list of all the users.
+   *
+   * @param ctx a Javalin HTTP context
+   */
+  public void addNewUser(Context ctx) {
+    User newUser = ctx.bodyValidator(User.class)
+      .check((usr) -> usr.name != null && usr.company.length() > 0) //Verify that the user has a name that is not blank
```
Comment (kklamberty) [inline server/src/main/java/umm3601/user/UserController.java]: We think that this is probably supposed to check for `usr.name.length()` not company

Comment (floogulinc) [reply server/src/main/java/umm3601/user/UserController.java]: Good catch!

**Comment by floogulinc** [comment]
Here's the update I posted in Slack:
- I have implemented all the endpoints on the server with mongo
  - This includes adding new users and deleting users (though this isnt used anywhere in the client yet)
- I have improved testing in the server with new mocks for the things that Javalin's Context uses so we can use the real class. The tests (i think they are more like integration tests at this point rather than unit tests) seem much more readable and understandable with some of the changes.
- I have brought mongo into the GitHub Actions workflows for the server and E2E. The server unit tests don't involve seeding the DB since they setup their own things for every test. The E2E one does seed the database with whatever is in the database/seed folder
- There is now a smalls script in the /database directory called mongoseed.sh this script automatically seeds any json files in /database/seed to their respective collections (based on their filenames) in the dev database by default (you can provide a different database as an argument).
- I have implemented an add user page and component in Angular using a lot of the code from the old lab 4 (so all of the validators and stuff are there) and updated it with a role selector and such.

What is left:
1. Fixing existing E2E tests
2. Writing some E2E tests for adding new users (checking if you can hit submit when you have entered bad data, checking what you entered VS what is on the user page you are redirected to, etc.)
3. Writing some unit tests for the add user component. These would probably be checking the logic of the validation and such.

**Review by kklamberty — COMMENTED** [review]
I'd like to understand more about why removing these tests makes sense. Were these tests replaced? Are these tests more appropriate as E2E tests?

**Pull request merged**